New gadget: tcpretrans

alban opened the pull request:
# New gadget: tcpretrans

The trace tcpretrans gadget traces TCP retransmissions by the kernel.

Based on bcc tool tcpretrans:
https://github.com/iovisor/bcc/blob/master/tools/tcpretrans_example.txt

It uses the socker enricher and the kube-ip-resolver like the tcpdrop gadget.

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/1423

## How to use

...

## Testing done

...

## TODO

- [ ] Complete implementation
- [ ] Documentation

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// This struct is defined according to the following format file:
+// /sys/kernel/tracing/events/tcp/tcp_retransmit_skb/format
+struct tcp_retransmit_skb_ctx {
+	/* The first 8 bytes are not allowed to read */
+	unsigned long pad;
+
+	void *skbaddr;
+	void *skaddr;
+	int state;
+	__u16 sport;
+	__u16 dport;
+	__u8 saddr[4];
+	__u8 daddr[4];
+	__u8 saddr_v6[16];
+	__u8 daddr_v6[16];
+};
```
`trace_event_raw_tcp_event_sk_skb` from vmlinux.h can be used instead -> https://github.com/torvalds/linux/blob/15fcdf6ae116d1e8da4ff76c6fd82514f5ea501b/include/trace/events/tcp.h#L98

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+	struct sock *sk = ctx->skaddr;
+
+	return __trace_tcp_retrans(ctx, sk, skb);
+
+	return 0;
```
nit: remove.

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+};
+
+// This struct is the same as struct tcphdr in vmlinux.h but with flags defined as single field instead of bitfield
+struct tcphdr_with_flags {
+	__be16 source;
+	__be16 dest;
+	__be32 seq;
+	__be32 ack_seq;
+	__u16 res1: 4;
+	__u16 doff: 4;
+	__u8 flags;
+	__be16 window;
+	__sum16 check;
+	__be16 urg_ptr;
+};
```
Not used?

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+	if (BPF_CORE_READ_BITFIELD_PROBED(sk, sk_protocol) != IPPROTO_TCP)
+		return 0;
+
+	struct inet_sock *sockp = (struct inet_sock *)sk;
```
In most of our gadgets we try to keep variable declarations at the top. (As done in the Linux kernel), perhaps we should try to follow that style.

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+	if (event.dport == 0)
+		return 0;
+
+	BPF_CORE_READ_INTO(&event.sport, sockp, inet_sport);
+	if (event.sport == 0)
+		return 0;
```
Out of curiosity. What's the different between doing it and `BPF_CORE_READ(sk, __sk_common.skc_num);` as done in other gadgets?

(I tried and it works the same, the only difference is that in this case the field is already in host byte order)

alban replied:
I am not sure to be honest.

mauriciovasquezbernal commented on pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c:
```diff
@@ -0,0 +1,155 @@
+	switch (event.af) {
+	case AF_INET:
+		BPF_CORE_READ_INTO(&event.daddr_v4, sk, __sk_common.skc_daddr);
+		if (event.daddr_v4 == 0)
+			return 0;
```
Do we need these checks?

alban replied:
I don't know. There is the same pattern in tcpdrop.bpf.c and tcptracer.bpf.c.

mauriciovasquezbernal reviewed (commented):
Very small nits. LGTM after those are fix. 

I suppose you'll handle documentation in another PR. Right?

alban commented:
Updated and added the documentation too.

Note that I run the commands from the documentation without #1576, so this explains why the socket enricher didn't work for nginx.

mauriciovasquezbernal reviewed (approved):
Some very small comments. I'm approving since my comments are just details.
